### ISSUE TITLE
Unresolvable reference when .git is a file

### ISSUE THREAD

**Issue by awwright** [isomorphic-git/isomorphic-git#735]
Git will sometimes configure the .git file to be a file, instead of a directory. This file has contents similar to:

```
gitdir: /home/awwright/repo.git/worktrees/work
```

This happens, among other cases, when a working tree is created with `git-worktree`. Here is a script showing the difference in behavior:

```
NAME=$(uuidgen).test
mkdir $NAME
pushd $NAME
git clone https://github.com/isomorphic-git/test.empty repo
git --git-dir=repo/.git worktree add work master

pushd work
echo '{"dependencies":{"isomorphic-git":"0.55.0"}}' > package.json
yarn install
echo
echo 'Git works:'
git log
echo
echo 'isomorphic-git bails:'
node -e 'require("isomorphic-git").log({depth:1, fs:require("fs"), dir:process.cwd(), ref:"master"}).then(console.log, console.error);'
popd

pushd repo
echo
echo 'but it works in the default working tree:'
node -e 'require("isomorphic-git").log({depth:1, fs:require("fs"), dir:process.cwd(), ref:"master"}).then(console.log, console.error);'
popd
popd
```

I would expect both calls to work, but only the second one does.

**Comment by wmhilton** [comment]
Wow, there goes `git` challenging my assumptions. `.git` can be a file!   😲 

**Short term**, here's a small tweak to your script to get it to work. Since you can specify the `gitdir` yourself, if you know where it is you can simply override it.

```diff
- node -e 'require("isomorphic-git").log({depth:1, fs:require("fs"), dir:process.cwd(), ref:"HEAD"}).then(console.log, console.error);'
+ node -e 'require("isomorphic-git").log({depth:1, fs:require("fs"), dir:process.cwd(), gitdir:require("path").resolve(process.cwd(), "../repo/.git"), ref:"HEAD"}).then(console.log, console.error);'
```

---

**Long term**: `isomorphic-git` should learn how to read and follow `.git` files. That's probably not _too_ hard, because it looks like in this case the file format is just:

    gitdir: /absolute/path/to/the/gitdir

and that appears to be it, according to [this Note](https://git-scm.com/docs/gitrepository-layout#_description). Not really a config file as much as a link.

This will require a little bit of refactoring! Currently the code simply appends `.git` to `dir` in the function default argument definition, like this:

```js
import { join } from '../utils/join.js'

function commit ({
  dir,
  gitdir = join(dir, '.git'),
  ...
}) {
  // do things with `dir` and `gitdir`
}
```

I think the simplest approach would be to extract that into a new function, like so:
 
```js
import { defaultGitDir } from '../utils/defaultGitDir.js'

function commit ({
  dir,
  gitdir = defaultGitDir(dir),
  ...
}) {
  // do things with `dir` and `gitdir`
}
```

and then in `../utils/defaultGitDir.js` add the ability to read `.git` files and use that pathname if it exists.

**Comment by wmhilton** [comment]
Ahhhh... hmm. Two more implementation notes for myself or whoever implements this feature.

1. default argument values can't be `await`ed. Since this will involve filesystem read, we'll have to move initializing it into the body of the function.
2. canonical `git` follows this "link" even if `GIT_DIR` is explicitly set. So _really_ we should leave it there and it should look more like this:

```
import { join } from '../utils/join.js'
import { resolveGitDir } from '../utils/resolveGitDir.js'

function commit ({
  dir,
  gitdir = join(dir, '.git'),
  ...
}) {
  gitdir = await resolveGitDir(gitdir)
  // do things with `dir` and `gitdir`
}
```

**Comment by awwright** [comment]
Thanks for the workaround @wmhilton!

**Comment by madscientist** [comment]
Lack of support for worktrees is biting me: trying to use some node facilities (Antora) that rely on isomorphic-git means I can't use them in worktrees.  And my workflow depends heavily on Git worktrees.  Hopefully someone will have some time to address this!

**Comment by seanpoulter** [comment]
Hi folks. I've just started with worktrees and have also run into this problem with Antora.

@wmhilton, would you welcome a PR? And to build some goodwill, want help going over the Dependabot PRs?

@madscientist and @awwright, are either of you interested in collaborating to drive this forward?

**Comment by jcubic** [comment]
@wmhilton is not active anymore. I'm right now taking care of maintaining the project. If you can create a reproduction of the issue that can be added to unit tests (that right now fail) you're more than welcome to fix the issue.

As for Dependabot PR, I think that we need to disable it somehow because where can't update some of the dependencies. But if you can update some deps and make a build pass you're also welcome to do so. I wanted some time ago to update all deps but a lot of things broke. Most of the deps that are security issues are for dev dependency that can't be updated unless the build system is also updated.

**Comment by seanpoulter** [comment]
Hello and thank you for the quick and clear reply @jcubic! I'm looking forward to working with you and hopefully driving things forward if I can find some time.

**Comment by mojavelinux** [comment]
I'm definitely keen on seeing built-in support for worktrees in isomorphic-git.

That said, I'm confused by the statement "run into this problem with Antora". Antora supports multiple worktrees. If you need help, I encourage you to join the project chat and ask questions there.

**Comment by seanpoulter** [comment]
Thanks @mojavelinux. Are you suggesting that we should be able to run the Antora CLI from a repo using git worktrees? What are you expecting it to support? From my limited understanding of Git worktrees:

* We can clone a repo like usual and add a worktree
* We can **bare** clone a repo and add a worktree

I've tried the second option using a bare repo and I end up with a similar error in as the original post.

**Comment by mojavelinux** [comment]
Yes, Antora supports git worktrees. See https://docs.antora.org/antora/latest/playbook/content-worktrees/ Please direct follow-up questions to the Antora project chat at https://chat.antora.org as we don't want to bog down this issue talking about Antora.

**Comment by gzaripov** [comment]
Hi @wmhilton !
I tried short-term solution and it doesn't work in case you want to resolve `HEAD` reference. Is happens because head references are different in `.git` and `.git/worktrees/worktree`. So to resolve `HEAD` you firstly should get ref from `.git/worktrees/worktree/HEAD` and then find this ref in `.git/refs/`.

**Comment by mojavelinux** [comment]
Here's the basic logic for resolving the repository starting from a git worktree, in case you find it useful: https://gitlab.com/antora/antora/-/blob/9bf33b35723f46fddaf742f84a116ca86c30c4b5/packages/content-aggregator/lib/aggregate-content.js#L1038-1054
